activerecord/lib/active_record/locking/optimistic.rb

```diff
@@ -152,6 +144,13 @@ def _clear_locking_column
           clear_attribute_change(self.class.locking_column)
         end
 
+        def _query_constraints_hash
+          return super unless locking_enabled?
+
+          locking_column = self.class.locking_column
+          super.merge(locking_column => _lock_value_for_database(locking_column))
```

Comment:
I think we could have mutated the hash rather than to use `merge`, but no big deal.